app/services/dataworks_mappers/sdr.rb

```diff
@@ -46,420 +56,508 @@ def perform_map
+    def titles
+      return [] if (title = source.dig('description', 'title', 0, 'value')).blank?
 
-      [
-        { title: Cocina::Models::Builders::TitleBuilder.full_title(element).first }
-      ]
+      [{ title: }]
     end
 
     # Druid is always present, but DOI may not be
     def identifiers
-      [
-        { identifier: source.externalIdentifier, identifier_type: 'DRUID' }
-      ].tap do |ids|
-        ids.push(doi) if doi.present?
-      end
-    end
+      ids = [{ identifier: source['externalIdentifier'], identifier_type: 'DRUID' }]
 
-    # DOI can be stored in two places, so check both
-    def doi
-      # If it's here, it is not in URL form
-      if source.identification.doi.present?
-        return {
-          identifier: source.identification.doi,
-          identifier_type: 'DOI'
-        }
+      if (doi_id = self.class.doi(source:)).present?
+        ids.push({ identifier: doi_id, identifier_type: 'DOI' })
       end
 
-      return if (doi = source.description.identifier.find { |id| id.type == 'DOI' }).nil?
-
-      # If it's here, it is in URL form
-      {
-        identifier: id_from_url(doi.value),
-        identifier_type: 'DOI'
-      }
+      ids
     end
 
-    # Strip the URL prefix from a DOI (or other identifier) and return just the ID
-    def id_from_url(id_url)
-      URI(id_url).path.delete_prefix('/')
+    def all_contributors
+      source.dig('description', 'contributor')&.map { |c| ContributorMapper.new(c) }&.compact || []
     end
 
     def creators
-      source.description.contributor.map do |c|
-        cocina_contributor_struct(c) if c.role.any? { |r| CREATOR_ROLES.include? r.value.downcase }
-      end.compact
+      all_contributors.filter(&:creator?).map(&:call)
     end
 
     def contributors
-      source.description.contributor.map do |c|
-        cocina_contributor_struct(c) unless c.role.any? { |r| CREATOR_ROLES.include? r.value.downcase }
-      end.compact
+      all_contributors.reject(&:creator?).map(&:call)
     end
 
     def publisher
-      source.description.event.flat_map(&:contributor).map do |c|
-        { name: c.name.first.value } if c.role.any? { |r| PUBLISHER_ROLES.include? r.value.downcase }
-      end
+      all_contributors.find(&:publisher?)&.call
     end
 
     # Use the earliest year from a publication date, or failing that, the
     # earliest year from a creation date. If no dates at all, fall back to the
     # year the source metadata was created.
     def publication_year
-      first_pub_year = date_values_by_type('Issued').map { |d| Date.edtf(d)&.year }.compact.min
-      created_year = date_values_by_type('Created').map { |d| Date.edtf(d)&.year }.compact.min
+      first_pub_year = date_years_by_type('Issued').min
+      created_year = date_years_by_type('Created').min
 
-      (first_pub_year || created_year || source.created.year).to_s
+      (first_pub_year || created_year || Time.zone.parse(source['created']).year).to_s
```

Comment:
Date parsing without error handling seems risky.

Reply:
This field is guaranteed to be present and always generated from a ruby `Date` object because it's an internal value...right? I don't think it's the same thing as a date from the metadata, which could be anything.

Also, we want this to blow up – if it does it means there's _nothing_ we can use for a publication date, which is a required field, so we need to look at the object to figure out what to do.